### PR TITLE
DDF-1819 Change SchematronValidationService to a managed service factory

### PR DESCRIPTION
**Pull request by spearskw**
@harrison-tarr 
@brendan-hofmann --> Hero
@ricklarsen

### CODE REVIEW

**Comment by brendan-hofmann** [comment]
Looks like you still need some documentation updates for this.

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java [hunk 1]
```diff
@@ -74,539 +69,164 @@
...
-                this.schematronSchemaFilename);
-
         this.suppressWarnings = suppressWarnings;
+    }
 
-        LOGGER.debug("EXITING: setSuppressWarnings");
+    public void setSchematronFileNames(ArrayList<String> schematronFileNames) {
+        this.schematronFileNames = new ArrayList<>();
+        for (String file : schematronFileNames) {
+            String fullpath = (file.startsWith("/")) ? file : SCHEMATRON_BASE_FOLDER + file;
```
Comment (brendan-hofmann) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: The assumption that all paths are either absolute or relative to `DDF_HOME/schematron` would be a good thing to document.

File: catalog/schematron/catalog-schematron-plugin/pom.xml [hunk 2]
```diff
@@ -69,7 +69,8 @@
                             ddf.services.schematron;version=${project.version}
                         </Export-Package>
                         <Embed-Dependency>
-                            catalog-core-api-impl
+                            catalog-core-api-impl,
+                            Saxon-HE
                         </Embed-Dependency>
                     </instructions>
                 </configuration>
```
Comment (brendan-hofmann) [inline catalog/schematron/catalog-schematron-plugin/pom.xml]: Can you update the code coverage? I'm sure it's higher than zero now. :smile_cat:

File: catalog/schematron/catalog-schematron-plugin/pom.xml [hunk 3]
```diff
@@ -69,7 +69,8 @@
                             ddf.services.schematron;version=${project.version}
                         </Export-Package>
                         <Embed-Dependency>
-                            catalog-core-api-impl
+                            catalog-core-api-impl,
```
Comment (stustison) [inline catalog/schematron/catalog-schematron-plugin/pom.xml]: Add !test scope to this embed.

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java [hunk 4]
```diff
@@ -74,539 +69,164 @@
...
-    /**
-     * Base directory in the bundle that all paths to resources will be relative to,
-     * e.g., "xyz" could be the bundle base directory and path to CVE files would be appended
-     * to this bundle base directory. A value of null indicates that "/" is the default
-     * base directory for the bundle.
-     */
-    private final String bundleBaseDir;
-
-    /** This class' logger */
+    private static final String SCHEMATRON_BASE_FOLDER = System.getProperty("ddf.home") + "/schematron/";
```
Comment (stustison) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: The "/" file separators may not work on Windows.

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java (hunk 1 above)
Comment (stustison) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: This won't work on Windows. Use Java Paths instead https://docs.oracle.com/javase/8/docs/api/java/nio/file/Paths.html

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java [hunk 5]
```diff
@@ -615,30 +235,27 @@ public void validate(Metacard metacard) throws ValidationException {
      */
     private static class SaxonErrorListener implements ErrorListener {
 
-        private String schematronSchemaFilename;
+        private ArrayList<String> schemaFiles;
```
Comment (stustison) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: List<String> this shouldn't require a specific implementation.

File: catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            interface="ddf.catalog.validation.MetacardValidator">
+        <cm:managed-component class="ddf.services.schematron.SchematronValidationService"
+                              init-method="init" destroy-method="destroy">
+
+            <property name="namespace" value=""/>
+            <property name="suppressWarnings" value="false"/>
+            <property name="schematronFileNames">
+                <list></list>
+            </property>
+            <cm:managed-properties persistent-id="" update-strategy="component-managed"
```
Comment (stustison) [inline catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Remove this.

File: catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/metatype/metatype.xml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    <AD name="Suppress Warnings" id="suppressWarnings" required="false" type="Boolean" default="false"
+        description="Don't show warnings generated by the schematron validator" />
+
+    <AD name="Schematron Files" id="schematronFileNames" required="true" type="String" cardinality="100"
+        description="Schematron files (*.sch) to be validated against" />
+
+  </OCD>
+
+  <Designate pid="ddf.services.schematron.SchematronValidationService" factoryPid="ddf.services.schematron.SchematronValidationService">
+    <Object ocdref="SchematronValidators" />
```
Comment (stustison) [inline catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/metatype/metatype.xml]: Just make the ocdref the same as the pids.

File: catalog/schematron/catalog-schematron-plugin/pom.xml (hunk 3 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/pom.xml]: !fixed

File: catalog/schematron/catalog-schematron-plugin/pom.xml (hunk 2 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/pom.xml]: !fixed

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java (hunk 1 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: !fixed

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java (hunk 5 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: !fixed

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java (hunk 4 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: !fixed

File: catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 6 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: !fixed

File: catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/metatype/metatype.xml (hunk 7 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/metatype/metatype.xml]: !fixed

**Comment by ricklarsen** [comment]
:+1:

File: catalog/docs/src/main/resources/IntContents.adoc [hunk 8]
```diff
@@ -1194,142 +1194,23 @@ Otherwise, caching is canceled if the user cancels the product download.
...
-By default, this suppress warnings option is unset (hence warnings result in invalid requests by default).
-
-Validation is executed per catalog entry in the ingest request.
-If multiple catalog entries are in the request, Schematron validation stops once a catalog entry is determined to be invalid.
-For example, if ten catalog entries are in a singe create ingest request and entry #4 is invalid, entries 5 through 10 will not even be validated. Schematron returns an invalid status after entry #4 is validated.
+Custom schematron rulesets can be used to validate metacard metadata.
+Multiple services can be created, and each service can have multiple rule sets associated with it.
+Namespaces are used to distinguish services.
+The root schematron files may be placed anywhere on the file system as long as they are configured with an absolute path.
+Any root schematron files with a relative path is assumed to be relative to DDF_HOME/schematron.
```
Comment (brendan-hofmann) [inline catalog/docs/src/main/resources/IntContents.adoc]: Typo: `files...is assumed` should be `files...are assumed`.

Also, we seem to have an inconsistent style on this, but I would prefer something like `<{branding}_HOME>`. @ricklarsen what are your thoughts?

File: catalog/schematron/catalog-schematron-plugin/pom.xml [hunk 9]
```diff
@@ -69,7 +58,9 @@
                             ddf.services.schematron;version=${project.version}
                         </Export-Package>
                         <Embed-Dependency>
-                            catalog-core-api-impl
+                            catalog-core-api-impl;scope=!test,
```
Comment (brendan-hofmann) [inline catalog/schematron/catalog-schematron-plugin/pom.xml]: Why the `scope=!test`?

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java [hunk 10]
```diff
@@ -74,537 +71,162 @@
...
-        LOGGER.debug("suppressWarnings = {}", suppressWarnings);
-
-        this.schematronSchemaFilename = schematronSchemaFilename;
-        this.bundleBaseDir = null;
-        this.suppressWarnings = suppressWarnings;
-        this.priority = DEFAULT_PRIORITY;
+    public void init() throws SchematronInitializationException {
+        if (transformerFactory == null) {
+            transformerFactory = TransformerFactory.newInstance(
+                    net.sf.saxon.TransformerFactoryImpl.class.getName(),
```
Comment (brendan-hofmann) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: Do you need the fully qualified name here? You're already importing this class.

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java [hunk 11]
```diff
@@ -74,537 +71,162 @@
...
-                            + this.schematronSchemaFilename, e);
-
-        // Would go here if an invalid .sch file was passed in
+            template = transformerFactory.newTemplates(stage3Output);
         } catch (Exception e) {
-            LOGGER.error("Couldn't create transfomer", e);
-            throw new SchematronInitializationException(
-                    "Error trying to create SchematronValidationService using sch file "
-                            + this.schematronSchemaFilename, e);
+            LOGGER.warn("Exception compiling schematron rules ", e);
```
Comment (brendan-hofmann) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: Why the change from error to warn? The validator is still going to blow up if compilation fails.

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java [hunk 12]
```diff
@@ -74,537 +71,162 @@
...
+        StringReader metadataReader = new StringReader(metadata);
+
+        SchematronReport report;
+        try {
+            Transformer transformer = validator.newTransformer();
+            DOMResult schematronResult = new DOMResult();
+            transformer.transform(new StreamSource(metadataReader), schematronResult);
+            report = new SvrlReport(schematronResult);
+        } catch (TransformerException e) {
+            throw new SchematronValidationException("Could not setup validator to perform validation.");
```
Comment (brendan-hofmann) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: It's fine to re-throw more descriptively, but I think the original exception should still be included in the re-throw.

File: catalog/docs/src/main/resources/IntContents.adoc (hunk 8 above)
Comment (ricklarsen) [inline catalog/docs/src/main/resources/IntContents.adoc]: I would prefer consistency, too, but the directory should actually be surrounded by backticks for monospace, which would escape the {branding} and render it literally.

Comment (brendan-hofmann) [inline catalog/docs/src/main/resources/IntContents.adoc]: I was afraid of that. I have some ideas on possible workarounds, but that's a different topic.

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java [hunk 13]
```diff
@@ -111,6 +118,57 @@ public static Result transform(Node nodeXml, TransformerProperties transformerPr
...
+        XMLStreamReader xmlStreamReader = null;
+
+        try (StringReader strReader = new StringReader(xml)) {
+            xmlStreamReader = xmlInputFactory.createXMLStreamReader(strReader);
+
+            while (xmlStreamReader.hasNext()) {
+                int event = xmlStreamReader.next();
+                if (event == XMLStreamConstants.START_ELEMENT) {
+                    rootNamespace = xmlStreamReader.getNamespaceURI();
+                    xmlStreamReader.close();
```
Comment (stustison) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java]: It is redundant to close the stream here and also in the finally. You should be able to remove this.

File: catalog/schematron/catalog-schematron-plugin/pom.xml (hunk 9 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/pom.xml]: Scott left a comment to add that, so I did.

File: catalog/docs/src/main/resources/IntContents.adoc (hunk 8 above)
Comment (spearskw) [inline catalog/docs/src/main/resources/IntContents.adoc]: !fixed

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java (hunk 10 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: !fixed

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java (hunk 11 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: !fixed

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java (hunk 12 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: !fixed

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java (hunk 13 above)
Comment (spearskw) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java]: !fixed

File: catalog/schematron/catalog-schematron-plugin/pom.xml (hunk 9 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/pom.xml]: !resolved

File: catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java [hunk 14]
```diff
@@ -73,538 +71,198 @@
...
+            String fullpath = Paths.get(filename).toString();
+            if (!Paths.get(filename).isAbsolute()) {
+                fullpath = Paths.get(SCHEMATRON_BASE_FOLDER, fullpath).toString();
             }
+            this.schematronFileNames.add(fullpath);
+        }
+        try {
+            updateValidators();
+        } catch (SchematronInitializationException e) {
+            // silently ignore
```
Comment (harrison-tarr) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: Why not log something here? I imagine silently failing would be a lot harder to debug

Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/java/ddf/services/schematron/SchematronValidationService.java]: I should add a comment as to why I did that actually, as well as logging I suppose.

When someone changes the schematron files in the admin UI, it calls this setter. I need to make new validators based on their changes, hence the call in the setter.

Now when the service is first initialized, it calls the setters first (hence this call will fail), then it calls init. I suppose try-catch really is the worst way to go here, I should check to see if it's initialized and then update if so.

Thanks!

**Comment by harrison-tarr** [comment]
:+1:

File: catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+<blueprint xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0"
+           xmlns:cm="http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.0.0">
+
+    <cm:managed-service-factory
+            id="ddf.services.schematron.SchematronValidationService"
+            factory-pid="ddf.services.schematron.SchematronValidationService"
+            interface="ddf.catalog.validation.MetacardValidator">
+        <cm:managed-component class="ddf.services.schematron.SchematronValidationService"
+                              init-method="init" destroy-method="destroy">
+            <cm:managed-properties persistent-id="" update-strategy="container-managed"/>
```
Comment (stustison) [inline catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I don't think you need this `cm:managed-properties`

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java [hunk 16]
```diff
@@ -111,6 +118,56 @@ public static Result transform(Node nodeXml, TransformerProperties transformerPr
...
+        if (xml == null) {
+            return null;
+        }
+
+        initializeXMLInputFactory();
+
+        XMLStreamReader xmlStreamReader = null;
+
+        try (StringReader strReader = new StringReader(xml)) {
+            xmlStreamReader = xmlInputFactory.createXMLStreamReader(strReader);
```
Comment (stustison) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java]: You can put this in the try with resources and remove the finally block.

File: catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 15 above)
Comment (spearskw) [inline catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I do need cm:managed-properties there, it breaks if I don't have it. Without that line it doesn't show up as a service when I go to system/console/services, so I'm guessing I need that line to create the services. I did try running without cm:managed-properties and the services wouldn't be created.

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java (hunk 16 above)
Comment (spearskw) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java]: XMLStreamReader doesn't extend AutoClosable, so I can't put it in the try-with-resources.

File: catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 15 above)
Comment (stustison) [inline catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Hm, OK!

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java (hunk 16 above)
Comment (stustison) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java]: Yep, you're right all good. For some reason when I tried it out, I wasn't getting any red lines, but now I just tried it again and did. Stupid intellij :)

**Comment by shaundmorris** [comment]
Looks like all comments have been addressed here so will merge.

File: catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 15 above)
Comment (harrison-tarr) [inline catalog/schematron/catalog-schematron-plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: fwiw, I've noticed the same thing. I had the same empty `cm:managed-properties` and it didn't work without it

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java (hunk 16 above)
Comment (shaundmorris) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java]: I was going to make the same comment but saw that XMLStreamReader didn't implement AutoClosable... Feature Request!

Comment (spearskw) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/XMLUtils.java]: I did actually try to do that at first. When I saw Scott's comment, I thought there might be some magic I missed, like maybe using woodstox instead. Alas, no luck. At least I was able to move the StringReader to the try-with-resources.

**Pull request merged**